### ISSUE TITLE
Nvidia GPU monitoring in container

### ISSUE THREAD

**Issue by D34DC3N73R** [netdata/netdata#4736]
I'm running the netdata/netdata docker container and I've installed nvidia-docker2 to give containers access to gpu, passing /dev/dri to the netdata container as well as environment variables `- NVIDIA_VISIBLE_DEVICES=all`, `- NVIDIA_DRIVER_CAPABILITIES=all`. The default runtime is `nvidia`. When I run `nvidia-smi` from within the container, I get `/usr/bin/nvidia-smi: No such file or directory`. However `ls /usr/bin/nvidia-smi` results in `/usr/bin/nvidia-smi` (it is present). This is on Ubuntu 16.04 with nvidia 410 driver. Any tips on how I can enable GPU monitoring in the netdata container? Below is my docker-compose.

```    netdata:
        container_name: netdata
        image: netdata/netdata
        ports:
            - 19999:19999
        restart: always
        network_mode: bridge
        environment:
            - TZ=America/Los_Angeles
            - PGID=999
            - NVIDIA_VISIBLE_DEVICES=all
            - NVIDIA_DRIVER_CAPABILITIES=all
        cap_add:
            - SYS_PTRACE
        security_opt:
            - seccomp:unconfined
        volumes:
            - /var/run/docker.sock:/var/run/docker.sock:ro
            - /proc:/host/proc:ro
            - /sys:/host/sys:ro
        devices:
            - /dev/dri:/dev/dri
        privileged: true

**Comment by paulfantom** [comment]
Netdata container doesn't ship `nvidia-smi`. So to enable it, you need to add this binary to a container. To do this you need to create your own container image. Your dockerfile should similar to this:
```
FROM netdata/netdata
RUN <commands_to_install_nvidia-smi_on_alpine_linux>
```

**Comment by D34DC3N73R** [comment]
It's my understanding that is taken care of with the nvidia-docker2 package. `docker run -e NVIDIA_VISIBLE_DEVICES=all --rm base/archlinux nvidia-smi` results as expected, outputting 

```Sun Nov 25 00:36:43 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.73       Driver Version: 410.73       CUDA Version: N/A      |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro P2000        Off  | 00000000:01:00.0 Off |                  N/A |
| 49%   39C    P0    18W /  75W |     89MiB /  5057MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+

**Comment by paulfantom** [comment]
I cannot reproduce it since i don't have any linux machine with nvidia card.

From what I see it looks more like a problem with nvidia-docker2 setup than with netdata. We don't ship `nvidia-smi` inside our image and nvidia-docker2 package makes some magic to add this into netdata container. But for some unidentified reason you cannot run it. Maybe it doesn't work with alpine images?

**Comment by D34DC3N73R** [comment]
That does seem to the case as the above command doesn't work with the Alpine image either. After searching through their GitHub issues it appears their drivers won't work with non glibc-based distributions. https://github.com/NVIDIA/nvidia-docker/issues/509

I haven't been able to build an Alpine image with Nvidia drivers, but it seems that would bloat the image to an unreasonable size, and it may not work anyways. I'm wondering if it'd be easier to create a netdata image based on minideb or BusyBox glibc etc.

**Comment by paulfantom** [comment]
It should be fairly easy to build netdata by using busybox instead of alpine. However currently we cannot switch netdata image to use busybox since [multiarch project](https://github.com/multiarch) doesn't have up to date busybox images (last was pushed 3 years ago).

We could look into using https://hub.docker.com/r/multiarch/debian-debootstrap/ image as a base, but this in turn drastically increases image size.

**Comment by D34DC3N73R** [comment]
I stumbled across https://github.com/frol/docker-alpine-glibc which does work with nvidia-docker2. I grabbed and edited the Dockerfile from [here](https://github.com/netdata/netdata/blob/master/docker/Dockerfile). When attempting to build netdata with it, I get the following error in the build process. I haven't really had a need to create my own image before this so any tips would be appreciated. 
```
Step 6/18 : RUN chmod +x netdata-installer.sh &&     sync && sleep 1 &&     ./netdata-installer.sh --dont-wait --dont-start-it
 ---> Running in 7995890b0ca4
chmod: netdata-installer.sh: No such file or directory
The command '/bin/sh -c chmod +x netdata-installer.sh &&     sync && sleep 1 &&     ./netdata-installer.sh --dont-wait --dont-start-it' returned a non-zero code: 1

**Comment by paulfantom** [comment]
Where did you execute it from? If you just changed netdata Dockerfile and it is still in `docker/` then you need to build you image with `docker build -t netdata-glibc --file=docker/Dockerfile ./` while being in top level directory of this repo.

Look here: https://github.com/netdata/netdata/blob/dbbe252e93e2b45e1e86e0cc73b754304a3c0b41/docker/build.sh#L33-L35

**Comment by D34DC3N73R** [comment]
I created a new directory `netdata-glibc` cd'ed into that and created the Dockerfile there. Then ran `docker build -t netdata-glibc .` I was under the impression the `.` told docker build to use the Dockerfile in the working directory.

**Comment by paulfantom** [comment]
No, `.` tells docker that your working directory is a dir where you currently are and docker assumes that Dockerfile is in this dir. Working directory is later used as a point of reference for things like COPY directive ([Like here](https://github.com/netdata/netdata/blob/master/docker/Dockerfile#L31)). That's why you usually see Dockerfile in tld.
Since we don't have Dockerfile in tld and we need files from repository for COPY directive you need to do one of the following (exclusive):
- modify Dockerfile
- execute build command as `docker build -t netdata-glibc --file=netdata-glibc/Dockerfile ./`

Second one is easier :smile:

**Comment by D34DC3N73R** [comment]
Thanks, I did get the image to build. At first I missed the second `FROM` in the Dockerfile, but I've now edited both instances to `frolvlad/alpine-glibc` and still don't see any reference to GPU monitoring.

**Comment by paulfantom** [comment]
Have you checked if you can run `nvidia-smi` inside that container?

**Comment by D34DC3N73R** [comment]
Yes that works now.
```
docker exec -i -t netdata nvidia-smi
Mon Nov 26 21:51:09 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.78       Driver Version: 410.78       CUDA Version: N/A      |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro P2000        Off  | 00000000:01:00.0 Off |                  N/A |
| 62%   33C    P0    18W /  75W |      0MiB /  5057MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

**Comment by paulfantom** [comment]
Did you enable `nvidia_smi` netdata module? By default it is disabled ([docs](https://docs.netdata.cloud/collectors/python.d.plugin/nvidia_smi/)).

**Comment by D34DC3N73R** [comment]
I see reference to NETDATA_USER_CONFIG_DIR and NETDATA_STOCK_CONFIG_DIR environment variables. Would they not be used here? Also looking at netdata.conf, I see 104 references to `nvidia` what exactly needs to be uncommented? If I map a volume, `.config/netdata/netdata.conf:/etc/netdata.conf` would that work?

**Comment by paulfantom** [comment]
Netdata installed in a container expects user configs in `/etc/netdata`. `nvidia_smi` is an external netdata module so to enable it you need to put configuration in `/etc/netdata/python.d.conf` file. All stock configuration files are located in `/usr/lib/netdata/conf.d`.

**Comment by D34DC3N73R** [comment]
I really appreciate all the help, but I'm just not wrapping my head around what file needs to be edited, and where it needs to be placed to enable GPU monitoring. 

I copied `/etc/netdata/python.d.conf` and uncommented `nvidia-smi=yes`. I then mapped the host copy to `/etc/netdata/python.d.conf` as a volume and started the container. After entering the container shell, `cat /etc/netdata/python.d.conf` shows the change, with `nvidia-smi=yes` uncommented. However, I still don't see any reference to GPU monitoring in the UI.

**Comment by D34DC3N73R** [comment]
As a side note, the glibc built image comes in at 105.6 MB while the netdata/netdata image is 110 MB. Using that alpine-glibc repo might be a nice way to enable GPU monitoring for other docker users in the future. I guess my next steps are to figure out how to automate the build of a glibc version of netdata upon new releases.

**Comment by paulfantom** [comment]
Main problem with `alpine-glibc` is that it is not supported by multiarch project, so we loose other platforms by switching to it.

**Comment by D34DC3N73R** [comment]
So what's the plan for enabling GPU stats in the container? Or is that not going to happen on the official image?

**Comment by D34DC3N73R** [comment]
Could something like this https://github.com/sgerrand/docker-glibc-builder be added on top of the multiarch alpine base? Maybe outputting to a different branch? netdata/netdata:glibc etc?

**Comment by paulfantom** [comment]
I need to check which other modules need glibc and if it is worth switching base image. But if we are going to switch base image it will be only to other "official" one so in this case it would be debian.

Other option is to use netdata/netdata as a base and add glibc on top of that.

**Comment by D34DC3N73R** [comment]
That actually works surprisingly well.
```
FROM netdata/netdata

RUN ALPINE_GLIBC_BASE_URL="https://github.com/sgerrand/alpine-pkg-glibc/releases/download" && \
    ALPINE_GLIBC_PACKAGE_VERSION="2.23-r3" && \
    ALPINE_GLIBC_BASE_PACKAGE_FILENAME="glibc-$ALPINE_GLIBC_PACKAGE_VERSION.apk" && \
    ALPINE_GLIBC_BIN_PACKAGE_FILENAME="glibc-bin-$ALPINE_GLIBC_PACKAGE_VERSION.apk" && \
    ALPINE_GLIBC_I18N_PACKAGE_FILENAME="glibc-i18n-$ALPINE_GLIBC_PACKAGE_VERSION.apk" && \
    apk add --no-cache --virtual=.build-dependencies wget ca-certificates && \
    wget \
        "https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub" \
        -O "/etc/apk/keys/sgerrand.rsa.pub" && \
    wget \
        "$ALPINE_GLIBC_BASE_URL/$ALPINE_GLIBC_PACKAGE_VERSION/$ALPINE_GLIBC_BASE_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BASE_URL/$ALPINE_GLIBC_PACKAGE_VERSION/$ALPINE_GLIBC_BIN_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BASE_URL/$ALPINE_GLIBC_PACKAGE_VERSION/$ALPINE_GLIBC_I18N_PACKAGE_FILENAME" && \
    apk add --no-cache \
        "$ALPINE_GLIBC_BASE_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BIN_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_I18N_PACKAGE_FILENAME" && \
    \
    rm "/etc/apk/keys/sgerrand.rsa.pub" && \
    /usr/glibc-compat/bin/localedef --force --inputfile POSIX --charmap UTF-8 C.UTF-8 || true && \
    echo "export LANG=C.UTF-8" > /etc/profile.d/locale.sh && \
    \
    apk del glibc-i18n && \
    \
    rm "/root/.wget-hsts" && \
    apk del .build-dependencies && \
    rm \
        "$ALPINE_GLIBC_BASE_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BIN_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_I18N_PACKAGE_FILENAME"

ENV LANG=C.UTF-8
```

**Comment by D34DC3N73R** [comment]
I set up an automated build with a trigger for netdata/netdata for anyone else who wants GPU support.  https://hub.docker.com/r/d34dc3n73r/netdata-glibc/

**Comment by prologic** [comment]
As discussed internally we _believe_ we can add appropriate documentation to our [docs](https://docs.netdata.cloud) however this isn't something we can make this automatic.

**Comment by Ferroin** [comment]
Relevant devices that need passed through should be something like `/dev/dri/card0` on most Linux systems.

**Comment by D34DC3N73R** [comment]
Nvidia cards take more than just passing /dev/dri. They also need drivers from the host system. The [nvidia-container-toolkit](https://github.com/NVIDIA/nvidia-docker) takes care of this but needs to be installed on the host system. Even with docker natively supporting GPUs in their runtime, this package still needs to be installed. For AMD cards to work, they'll have to make a similar package. The unfortunate thing about the nvidia package is it only works with glibc distributions. That's the entire reason I added the glibc library to the netdata/netdata base.

**Comment by prologic** [comment]
@D34DC3N73R Since you clearly know a fair bit about NVIDIA GPU(s) here and how to run them inside Docker containers (_none of us currently have access to such things at the moment_); Would you be willing to put up a PR that documents this? I don't think we can support this type of setup more than documenting how it can be done. Thoughts?

**Comment by D34DC3N73R** [comment]
I'm afraid I don't know of a solution that allows netdata to support GPUs in docker containers other than adding the [glibc package](https://github.com/sgerrand/alpine-pkg-glibc) to the official image. That's essentially the only thing [my netdata image](https://github.com/D34DC3N73R/netdata-glibc) does, other than enabling the nvidia-smi module in python.d.conf by default. 

After the glibc package is added, it's fairly trivial to install nvidia drivers and the [nvidia container toolkit](https://github.com/NVIDIA/nvidia-docker).

For some background, native gpu support in docker (v19.03+) is a recent feature but still requires both nvidia drivers and the nvidia container toolkit. Native gpu support has not yet come to docker-compose, so the nvidia docker 2 package, nvidia drivers, (and modifying /etc/docker/daemon.json to make nvidia the default runtime in docker-compose v3+), is still required. If you'd like me to expand on any of this, I'd be happy to.

**Comment by lgg** [comment]
**TL;DR:**

## Solution

* use this image: https://github.com/D34DC3N73R/netdata-glibc instead of original image of netdata
  * details about how it works: https://github.com/netdata/netdata/issues/4736

## Screenshots

![image](https://user-images.githubusercontent.com/12018713/159197540-5eee51e9-3c5f-4b0b-815c-88dfe6641267.png)

![image](https://user-images.githubusercontent.com/12018713/159197559-c039ac3b-5991-4551-86f0-45d337e5c534.png)

check: https://gist.github.com/lgg/f76d0124784b62baffac6ffbbb065e60

**Comment by lgg** [comment]
@Ferroin are you okay with it - if I will add info about using netdata with nvidia-smi in docker to docs (https://github.com/netdata/netdata/blob/master/collectors/python.d.plugin/nvidia_smi/README.md) referring to external image's repo: https://github.com/D34DC3N73R/netdata-glibc? 

Or this won't be merged?

**Comment by Ferroin** [comment]
@lgg Provided there is an explicit mention that it’s provided by a third party (I want to make sure users know this ahead of time instead of being caught completely off guard when we ask them to reproduce an issue unrelated to this with an official image) I see no reason not to mention it in the documentation.

**Comment by D34DC3N73R** [comment]
@Ferroin @prologic I've revisited this and netdata docker images will work with nvidia-container-toolkit by including `gcompat` in netdata/base. This still adds a glibc compatibility layer, but with a standard alpine package which I believe is multi-arch.

Example Dockerfile
```
FROM netdata/netdata:stable
RUN apk add --no-cache gcompat
```

example docker-compose.yml
```
  netdata:
    image: netdata-glibc:gcompat
    container_name: netdata
    hostname: netdata.xxxxxxx.xxx
    ports:
      - 19999:19999
    restart: unless-stopped
    depends_on:
      - proxy
    cap_add:
      - SYS_PTRACE
    security_opt:
      - apparmor:unconfined
    environment:
      - DOCKER_HOST=proxy:2375
      - NETDATA_CLAIM_TOKEN=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
      - NETDATA_CLAIM_URL=https://app.netdata.cloud
      - NETDATA_CLAIM_ROOMS=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
    volumes:
      - $HOME/.config/netdata/config:/etc/netdata
      - $HOME/.config/netdata/lib:/var/lib/netdata
      - $HOME/.config/netdata/cache:/var/lib/cache/netdata
      - /etc/passwd:/host/etc/passwd:ro
      - /etc/group:/host/etc/group:ro
      - /proc:/host/proc:ro
      - /sys:/host/sys:ro
      - /etc/os-release:/host/etc/os-release:ro
    deploy:
      resources:
        reservations:
          devices:
          - driver: nvidia
            count: all
            capabilities: [gpu]

  proxy:
    container_name: proxy
    image: tecnativa/docker-socket-proxy
    restart: unless-stopped
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    environment:
      - CONTAINERS=1
```

result
```
docker exec -it netdata nvidia-smi
Thu Apr 13 19:15:45 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: N/A      |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Quadro P2000        Off  | 00000000:09:00.0 Off |                  N/A |
| 48%   38C    P0    18W /  75W |   3406MiB /  5120MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```
and after enabling nvidia-smi python plugin
<img width="1614" alt="image" src="https://user-images.githubusercontent.com/9123670/231861573-b0d7cf95-b450-4904-81e5-2363a9135496.png">

I'd be happy to answer and questions or help with documentation if needed.

**Comment by D34DC3N73R** [comment]
@fzyzcjy yes, that is correct. I'll also mention that gcompat is no longer required as netdata has switched to a debian base image. This works with the stock netdata image with the nvidia container toolkit and no additional software.

**Comment by ilyam8** [comment]
I updated docs in #17167.

My understanding is that for 

- `docker run`: adding `--gpus 'all,capabilities=utility'`
- `docker compose`: adding netdata netdata service

   ```yaml
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: all
              capabilities: [gpu]
   ```

Correct, @D34DC3N73R?

I didn't get what `capability: [gpu]` is. It is different from docker run gpus capabilities, right? Because there is no `gpu` [in the list](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/docker-specialized.html#driver-capabilities).

**Comment by D34DC3N73R** [comment]
@ilyam8 For docker run `capabilities=utility` is only required if you wish to restrict the capability of the card in the container. 
`--gpus all` would suffice.

For docker compose, I believe `capabilities: [gpu]` is a placeholder / default flag which is similar to not specifying any capabilities in the docker run command. It is there because the command fails if it is not filled in.
